### PR TITLE
Foundation: port NSHost to Windows

### PR DESCRIPTION
**Pull request by compnerd**
This ports the NSHost functionality to Windows.

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
CC: @parkera @phausler @millenomi @jakepetroules

**Comment by compnerd** [comment]
@swift-ci please test

File: Foundation/Host.swift [hunk 1]
```diff
@@ -33,6 +33,18 @@ open class Host: NSObject {
...
     
     static internal func currentHostName() -> String {
+#if os(Windows)
+      var dwLength: DWORD = 0
+      GetComputerNameExA(ComputerNameDnsHostname, nil, &dwLength)
+
+      let hostname: UnsafeMutablePointer<Int8>? =
+          UnsafeMutableBufferPointer<Int8>
+              .allocate(capacity: Int(dwLength + 1))
+              .baseAddress
```
Comment (spevans) [inline Foundation/Host.swift]: In scl-foundation source, long lines arent wrapped.

File: Foundation/Host.swift [hunk 2]
```diff
@@ -33,6 +33,18 @@ open class Host: NSObject {
...
+      defer { hostname?.deallocate() }
+      GetComputerNameExA(ComputerNameDnsHostname, hostname, &dwLength)
+      return String(cString: hostname!)
```
Comment (spevans) [inline Foundation/Host.swift]: Can a check be added to return `"localhost"` if `hostname` is empty?

File: Foundation/Host.swift [hunk 3]
```diff
@@ -33,6 +33,18 @@ open class Host: NSObject {
     }
     
     static internal func currentHostName() -> String {
+#if os(Windows)
+      var dwLength: DWORD = 0
+      GetComputerNameExA(ComputerNameDnsHostname, nil, &dwLength)
+
+      let hostname: UnsafeMutablePointer<Int8>? =
```
Comment (spevans) [inline Foundation/Host.swift]: Does `hostname` need to be an optional?

File: Foundation/Host.swift (hunk 1 above)
Comment (compnerd) [reply Foundation/Host.swift]: I really wish that `swift-format` was setup.  I'll join the lines if thats the style here.

File: Foundation/Host.swift (hunk 2 above)
Comment (compnerd) [reply Foundation/Host.swift]: I don't think that is possible (that is, `hostname` being empty).  We know the size of the host name due to the query above, so we should have plenty of space.  I suppose that it is possible to hit the race condition with the TOCTOU here ... but that does feel like an edge case.

File: Foundation/Host.swift (hunk 3 above)
Comment (compnerd) [reply Foundation/Host.swift]: hmm, I suppose not anymore, since it is being assigned a value immediately.

File: Foundation/Host.swift (hunk 2 above)
Comment (spevans) [reply Foundation/Host.swift]: I think its worth adding something like
```
guard dwLength > 0 && GetComputerNameExA(..) else {
  return "localhost"
}
```
just to cover the edge case for completeness, even if its unlikely to get used.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test
